python/ray/data/_internal/numpy_support.py

```diff
@@ -38,6 +39,28 @@ def validate_numpy_batch(batch: Union[Dict[str, np.ndarray], Dict[str, list]]) -
             f"a numpy batch to a block, got: {type(batch)} "
             f"({_truncated_repr(batch)})"
         )
+ 
+def _detect_highest_datetime_precision_dtype(datetime_list: List[datetime]) -> str:
+    highest_precision = 'datetime64[D]'  # Start with day precision
+    
+    for dt in datetime_list:
+        if dt.microsecond != 0:
+            highest_precision = 'datetime64[ns]'
+            break
+        elif dt.second != 0:
+            highest_precision = 'datetime64[s]'
+        elif dt.minute != 0:
+            highest_precision = 'datetime64[m]'
+        elif dt.hour != 0:
+            highest_precision = 'datetime64[h]'
```

Comment:
For my own understanding,  is this list of precisions exhaustive?

Reply:
The full list is here: https://numpy.org/doc/stable/reference/arrays.datetime.html#datetime-units. 

Not including lower precision units (W, M, Y) is not a problem unless people need to represent times so distance that the extra range is needed (for context day precision allows for linux epoch +/- 2.5e16 years).

For higher precision units (ns, ps, fs, as), the precision of python datetime is microseconds so I don't think we need to support those in this function.